src/Query/Filter/BooleanField.ts

```diff
@@ -0,0 +1,154 @@
+import { PostfixExpression, parse as boonParse } from 'boon-js';
+
+import { parseFilter } from '../FilterParser';
+import type { Task } from '../../Task';
+import { Field } from './Field';
+import { Filter, FilterOrErrorMessage } from './Filter';
+
+/**
+ * BooleanField is a 'container' field type that parses a high-level filtering query of
+ * the format --
+ *    (filter1) AND ((filter2) OR (filter3))
+ * The filters can be mixed and matched with any boolean operators as long as the individual filters are
+ * wrapped in either paranthesis or quotes -- (filter1) or "filter1".
+ * What happens internally is that when the boolean field is asked to create a filter, it parses the boolean
+ * query into a logical postfix expression (https://en.wikipedia.org/wiki/Reverse_Polish_notation),
+ * with the individual filter components as "identifier" tokens.
+ * These identifiers have an associated actual Filter (which is cached during the query parsing).
+ * The returned Filter of the whole boolean query is eventually a function that for each Task object,
+ * evaluates the complete postfix expression by going through the individual filters and then resolving
+ * the expression into a single boolean entity.
+ */
+export class BooleanField extends Field {
+    private static readonly basicBooleanRegexp = /.*(AND|OR|XOR|NOT).*/g;
```

Comment:
I wondered if this `basicBooleanRegexp` was a bit too relaxed.

## Prior to this PR

With this input:

````
# Boolean Queries

- [ ] I am a task with AND in description

---

```tasks
path includes boolean
description includes AND
```
````

I get this result:

![image](https://user-images.githubusercontent.com/4840096/177216806-c666a7fa-86d2-41ea-962e-a7281e67bda9.png)

## After this PR

With the same input, I get:

`Tasks query: malformed boolean query -- Invalid token (check the documentation for guidelines)`

![image](https://user-images.githubusercontent.com/4840096/177216733-5ac7fdf1-0cfe-4d24-b4a7-2d992b9bd3cc.png)

Reply:
I think that this may mean that if there is an error in parsing the query, `canCreateFilterForLine()` should return false.

Or, another idea might be to move the Boolean filter to the end of the 'chain of responsibly' list of constructors, so that any other filters are matched first... In this case, description would have triggered first, and it would never have reached the Boolean one.

Reply:
1. I can move more logic into `canCreateFilterForLine` and actually create the postfix expression there, but this would mean that if the user actually tried to create a boolean query and there's a syntax error, he will not get a useful error message (like you had when `XOR` was not implemented).
2. One thing that is very doable is to slightly improve the regex in `canCreateFilterForLine`, that would enforce quotes or parenthesis next to the operator that it looks for. It's not bullet-proof at all but will filter out the case you found.
3. My original reason for putting `BooleanField` first was that a boolean line can contain any other field text, so I was concerned that some existing or future field types may interpret things like `(field text` as their own. But from sampling a few other fields, I see now that they're quite strict about accepting a text and enforcing a beginning of line or the full string. I still need to review all of them to be sure though (otherwise it will be a very hard-to-catch bug).

Reply:
@esm7 Thank you!
1. Totally agree.
2. Seems like a good idea for a next step.
3. Good point, but I am uncomfortable with 'need to review all of them to be sure though', as it doesn't guard against future errors. This is an ideal candidate for adding some test code to detect if that situation ever happens in future. Here is some pseudocode:

```typescript
for each creator in fieldCreators {
    if (BooleanFilter) continue;
    get the creator regex
    if (regex does not being with ^) log a test error
}
```

Reply:
So, I just pushed a fix with the following:

1. Improved the regex in `BooleanField` to be more selective about what it accepts (this by itself already fixes the issue pointed out, but is not strong enough).
2. Moved `BooleanField` to the end of `fieldCreators` (this also fixes the issue by itself).
3. Added test cases that catch the original issue (+saw them fail before the above and succeed now).
4. Added a new comprehensive "should not confuse a boolean query for any other single field" test, that for each filter in the long list in `Query.test.ts`, composes a `Query` object and verifies it works correctly. This way any filter that is added to that list in the future will be verified to not "steal" a boolean query.

Reply:
Thank you! I'll review as soon as I can - likely tomorrow...

Reply:
It looks good. Thank you!.

There's a few more tests I would like to add, but they should not stop this being merged.
Resolving conversation.